<a href="https://colab.research.google.com/github/YaPineiro/Fin420/blob/main/YP_week_1_session_0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 1 Session 0

# Do Higher Levels of Institutional Ownership Increase Management Effectiveness?

Imagine you arrive at work on a Monday morning and this is the question your boss asks you.  Theory would prompt you to answer 'yes', however you'll want some data to back up your answer.  In this session we'll gather some relevant data and provide an answer.

Load the Pandas library.

In [1]:
import pandas as pd

### Get institutional ownership data.

The URL of an Excel spreadsheet which contains the amount of Institutional and Insider ownership by industry.

In [2]:
inst_own_url = "http://www.stern.nyu.edu/~adamodar/pc/datasets/inshold.xls"

Below we are going to import the data from the url using the `read_excel` method.  This attempts to extablish a SSL connection, and as of writing this the required certificates were not on the server.  The following code tells python to ignore the lack of valid certs.  It is not the safest approach, though this is the benefit of working on a notebook in the cloud.

In [3]:
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

Note the data starts on row 8, so we skip the first 7 rows when creating the Pandas `DataFrame`.  Also, the data is in sheet 2 (which in Python is 1).

In [4]:
inst_own = pd.read_excel(inst_own_url, skiprows = 7, sheet_name=1)

### Inspecting Data

Often the first step will be to inspect the data.  We want to pay particular attention to missing data and the type of each variable (column).  We can first view the data with:

In [5]:
inst_own

Industry Name  Number of Firms  CEO Holding  \
0                         Advertising               54     0.099896   
1                   Aerospace/Defense               67     0.034654   
2                       Air Transport               24     0.030976   
3                             Apparel               37     0.053675   
4                        Auto & Truck               34     0.096935   
..                                ...              ...          ...   
91                           Trucking               24     0.054088   
92                  Utility (General)               14     0.001355   
93                    Utility (Water)               15     0.005923   
94                       Total Market             6062     0.048202   
95  Total Market (without financials)             4935     0.048431   

    Corporate Holdings  Institutional Holdings  Insider Holdings  
0             0.152666                0.352435          0.233282  
1             0.180971                0.528903          0.088726  
2             0.261742                0.418113          0.137485  
3             0.108202                0.514791          0.170123  
4             0.195019                0.230086          0.174605  
..                 ...                     ...               ...  
91            0.340488                0.583142          0.168315  
92            0.000000                0.857136          0.004635  
93            0.276800                0.556169          0.059443  
94            0.141806                0.484652          0.127259  
95            0.141046                0.496171          0.126508  

[96 rows x 6 columns]

EDIT:  There is no longer a blank column, so no need to run the next line of code.  And it looks like we have a blank column (Unnamed: 5).  We can remove it with:

In [6]:
inst_own.drop(columns=["Unnamed: 5"], inplace=True)

KeyError: "['Unnamed: 5'] not found in axis"

Now let's see how many rows and columns we have (the shape of our data frame):

In [7]:
inst_own.shape

(96, 6)

Or just get the rows:

In [8]:
inst_own.shape[0]

96

And the type of each column:

In [9]:
inst_own.dtypes

Industry Name              object
Number of Firms             int64
CEO Holding               float64
Corporate Holdings        float64
Institutional Holdings    float64
Insider Holdings          float64
dtype: object

Or we can also use `info()`.  This prints a collection of useful information about a `DataFrame`.

In [10]:
inst_own.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Industry Name           96 non-null     object 
 1   Number of Firms         96 non-null     int64  
 2   CEO Holding             96 non-null     float64
 3   Corporate Holdings      96 non-null     float64
 4   Institutional Holdings  96 non-null     float64
 5   Insider Holdings        96 non-null     float64
dtypes: float64(4), int64(1), object(1)
memory usage: 4.6+ KB


And so the Number of Firms is an integer type. With the remaining columns being floats (numbers with decimal portions).  The `Object` for Industry Name in Pandas means it is a string. These columns have been imported correctly.  

A common problem when importing numeric column, particularly in large data sets, is there may be a character in one cell.  In this case the whole column may be imported as strings ("123" instead of 123).

### Example: Importing Data with Error

In [11]:
inst_own_with_error = pd.read_excel("https://github.com/FinancialMarkets/industry_instown_with_data_error_to_clean/blob/master/which_column_has_bad_data.xls?raw=true", skiprows = 7)
inst_own_with_error.drop(columns=["Unnamed: 5"], inplace=True)

In [12]:
inst_own_with_error.dtypes

Industry Name              object
Number of Firms             int64
CEO Holding               float64
Institutional Holdings     object
Insider Holdings          float64
dtype: object

In [13]:
inst_own_with_error

Industry Name  Number of Firms  CEO Holding  \
0                         Advertising               47     0.050409   
1                   Aerospace/Defense               77     0.030481   
2                       Air Transport               18     0.017713   
3                             Apparel               51     0.091342   
4                        Auto & Truck               13     0.069599   
..                                ...              ...          ...   
91                           Trucking               33     0.020479   
92                  Utility (General)               16     0.001173   
93                    Utility (Water)               17     0.021316   
94                       Total Market             7053     0.047442   
95  Total Market (without financials)             5878     0.049319   

   Institutional Holdings  Insider Holdings  
0                0.375451          0.161023  
1                0.571001          0.099429  
2                0.730171          0.042613  
3                0.505583          0.180427  
4                0.606275          0.095710  
..                    ...               ...  
91                0.55906          0.157253  
92               0.803944          0.005601  
93               0.499558          0.052463  
94               0.485737          0.129176  
95               0.500262          0.129637  

[96 rows x 5 columns]

Now we see `Institutional Holdings` is a string, and not a number.  Let's take a look and try to figure out where the error is:

In [14]:
pd.set_option('display.max_rows', 1000)
inst_own_with_error["Institutional Holdings"]

0          0.375451
1          0.571001
2          0.730171
3          0.505583
4          0.606275
5          0.599107
6          0.584515
7          0.354313
8           0.45555
9          0.261087
10         0.587724
11         0.656569
12         0.685724
13         0.524726
14          0.57085
15          0.37529
16          0.79305
17         0.574201
18         0.362273
19         0.549416
20         0.457172
21         0.727958
22         0.540199
23         0.361361
24         0.272346
25    5g3.92631818%
26         0.406755
27         0.281636
28         0.449221
29         0.676232
30         0.439051
31         0.455902
32         0.440725
33         0.388296
34         0.497551
35          0.62787
36         0.425788
37         0.342021
38         0.461281
39         0.491228
40         0.505673
41         0.768542
42         0.537949
43         0.629228
44         0.338093
45         0.621787
46         0.733088
47         0.657938
48         0.568681
49         0.470696
50         0.673124
51         0.208205
52         0.602455
53           0.7296
54         0.377124
55         0.565526
56          0.52961
57         0.694487
58          0.75416
59         0.709458
60          0.20258
61          0.56615
62         0.765564
63         0.308433
64         0.530997
65          0.34201
66         0.518088
67          0.89815
68         0.597804
69         0.741948
70         0.775873
71          0.60425
72         0.728622
73         0.477076
74         0.423614
75         0.674289
76         0.467243
77         0.588169
78         0.640928
79         0.384601
80         0.781213
81         0.366207
82         0.466025
83         0.480943
84         0.667756
85          0.49328
86         0.460804
87          0.38819
88         0.416315
89         0.824517
90          0.63909
91          0.55906
92         0.803944
93         0.499558
94         0.485737
95         0.500262
Name: Institutional Holdings, dtype: object

We see in row 25 there is a letter `g` in the number.  We have two choices at this point, we can (1) remove the row, or (2) if we are sure the number is otherwise correct we can remove the `g`.

In [15]:
inst_own_with_error.index

RangeIndex(start=0, stop=96, step=1)

#### Drop a Row

We can drop a row with the following.  Note the `inplace=True` is commented.  We want the error to stay in the DataFrame for the second posssible solution.  

In [16]:
inst_own_with_error.drop(axis=0, index=25) #, inplace=True)

Industry Name  Number of Firms  CEO Holding  \
0                              Advertising               47     0.050409   
1                        Aerospace/Defense               77     0.030481   
2                            Air Transport               18     0.017713   
3                                  Apparel               51     0.091342   
4                             Auto & Truck               13     0.069599   
5                               Auto Parts               46     0.029056   
6                      Bank (Money Center)                7     0.001516   
7                         Banks (Regional)              611     0.017899   
8                     Beverage (Alcoholic)               21     0.016225   
9                          Beverage (Soft)               34     0.040421   
10                            Broadcasting               27     0.075388   
11          Brokerage & Investment Banking               39     0.045112   
12                      Building Materials               42     0.018934   
13            Business & Consumer Services              165     0.072431   
14                                Cable TV               14     0.005457   
15                        Chemical (Basic)               43     0.085014   
16                  Chemical (Diversified)                6     0.014690   
17                    Chemical (Specialty)               94     0.020830   
18                   Coal & Related Energy               22     0.028429   
19                       Computer Services              106     0.073701   
20                   Computers/Peripherals               48     0.036865   
21                   Construction Supplies               44     0.037227   
22                             Diversified               23     0.022122   
23                   Drugs (Biotechnology)              503     0.033630   
24                  Drugs (Pharmaceutical)              267     0.048045   
26                    Electrical Equipment              113     0.050020   
27         Electronics (Consumer & Office)               20     0.086656   
28                   Electronics (General)              153     0.039852   
29                Engineering/Construction               54     0.069114   
30                           Entertainment              107     0.092871   
31          Environmental & Waste Services               82     0.074181   
32                     Farming/Agriculture               31     0.046266   
33  Financial Svcs. (Non-bank & Insurance)              232     0.037064   
34                         Food Processing               88     0.050342   
35                        Food Wholesalers               17     0.048782   
36                   Furn/Home Furnishings               35     0.059996   
37                Green & Renewable Energy               22     0.098471   
38                     Healthcare Products              242     0.041659   
39             Healthcare Support Services              128     0.065537   
40    Heathcare Information and Technology              129     0.074792   
41                            Homebuilding               32     0.075656   
42         Hospitals/Healthcare Facilities               36     0.042355   
43                            Hotel/Gaming               65     0.052368   
44                      Household Products              127     0.126213   
45                    Information Services               69     0.061668   
46                     Insurance (General)               19     0.021413   
47                        Insurance (Life)               24     0.007723   
48                   Insurance (Prop/Cas.)               51     0.041256   
49          Investments & Asset Management              192     0.107877   
50                               Machinery              120     0.026871   
51                         Metals & Mining               92     0.131931   
52             Office Equipment & Services               22     0.035822   
5

In [17]:
inst_own_with_error.dtypes

Industry Name              object
Number of Firms             int64
CEO Holding               float64
Institutional Holdings     object
Insider Holdings          float64
dtype: object

#### Changing the Data in Row 25 of Institutional Holdings

In [18]:
inst_own_with_error["Institutional Holdings"].loc[25] = 0.539263
#inst_own_with_error["Institutional Holdings"].loc[25]

<ipython-input-18-4c4355c97a23>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  inst_own_with_error["Institutional Holdings"].loc[25] = 0.539263


In [19]:
inst_own_with_error["Institutional Holdings"] = pd.to_numeric(inst_own_with_error["Institutional Holdings"])

And we see below Institutional Holdings is a float.

In [20]:
inst_own_with_error.dtypes

Industry Name              object
Number of Firms             int64
CEO Holding               float64
Institutional Holdings    float64
Insider Holdings          float64
dtype: object

### Selecting Columns

In [21]:
pd.set_option('display.max_rows', 10)
inst_own["Industry Name"]

0                           Advertising
1                     Aerospace/Defense
2                         Air Transport
3                               Apparel
4                          Auto & Truck
                    ...                
91                             Trucking
92                    Utility (General)
93                      Utility (Water)
94                         Total Market
95    Total Market (without financials)
Name: Industry Name, Length: 96, dtype: object

### Adding Columns

Lets say we want to create a column which is the sum of Institutional and Insider Holdings.  Let's call it `II_Holdings`.

In [22]:
inst_own["II_Holdings"] = inst_own["Institutional Holdings"] + inst_own["Insider Holdings"]
inst_own

Industry Name  Number of Firms  CEO Holding  \
0                         Advertising               54     0.099896   
1                   Aerospace/Defense               67     0.034654   
2                       Air Transport               24     0.030976   
3                             Apparel               37     0.053675   
4                        Auto & Truck               34     0.096935   
..                                ...              ...          ...   
91                           Trucking               24     0.054088   
92                  Utility (General)               14     0.001355   
93                    Utility (Water)               15     0.005923   
94                       Total Market             6062     0.048202   
95  Total Market (without financials)             4935     0.048431   

    Corporate Holdings  Institutional Holdings  Insider Holdings  II_Holdings  
0             0.152666                0.352435          0.233282     0.585717  
1             0.180971                0.528903          0.088726     0.617629  
2             0.261742                0.418113          0.137485     0.555598  
3             0.108202                0.514791          0.170123     0.684914  
4             0.195019                0.230086          0.174605     0.404692  
..                 ...                     ...               ...          ...  
91            0.340488                0.583142          0.168315     0.751457  
92            0.000000                0.857136          0.004635     0.861771  
93            0.276800                0.556169          0.059443     0.615613  
94            0.141806                0.484652          0.127259     0.611911  
95            0.141046                0.496171          0.126508     0.622678  

[96 rows x 7 columns]

# Which Industries have the Highest Levels of Institutional Ownership?

We can sort the `DataFrame` on `Institutional Ownership`, and print the first 10 rows, with:

In [23]:
inst_own.sort_values("Institutional Holdings", ascending=False).head(10)

Industry Name  Number of Firms  CEO Holding  \
67                 Reinsurance                1     0.000350   
92           Utility (General)               14     0.001355   
12          Building Materials               39     0.018672   
74              Retail (REITs)               28     0.005538   
90  Transportation (Railroads)                4     0.001335   
59                       Power               48     0.007155   
41                Homebuilding               30     0.079001   
5                   Auto Parts               33     0.051528   
78         Semiconductor Equip               30     0.017968   
80                        Shoe               12     0.009449   

    Corporate Holdings  Institutional Holdings  Insider Holdings  II_Holdings  
67            0.000000                0.965300          0.004050     0.969350  
92            0.000000                0.857136          0.004635     0.861771  
12            0.102658                0.818994          0.052210     0.871204  
74            0.027237                0.770551          0.027786     0.798337  
90            0.000010                0.768775          0.003240     0.772015  
59            0.058685                0.764030          0.023714     0.787744  
41            0.069467                0.756331          0.153598     0.909929  
5             0.133344                0.711526          0.093646     0.805172  
78            0.052257                0.701565          0.086590     0.788155  
80            0.087100                0.698034          0.055155     0.753189

We can also use `describe()` to get a fell for the data.

In [24]:
inst_own.describe()

Number of Firms  CEO Holding  Corporate Holdings  \
count        96.000000    96.000000           96.000000   
mean        177.697917     0.048738            0.150023   
std         789.957589     0.033600            0.100104   
min           1.000000     0.000350            0.000000   
25%          18.000000     0.019231            0.086743   
50%          32.500000     0.049144            0.129665   
75%          66.250000     0.065548            0.193874   
max        6062.000000     0.166336            0.583400   

       Institutional Holdings  Insider Holdings  II_Holdings  
count               96.000000         96.000000    96.000000  
mean                 0.535886          0.125710     0.661597  
std                  0.139598          0.059546     0.117247  
min                  0.230086          0.003240     0.404692  
25%                  0.428347          0.089808     0.573631  
50%                  0.532624          0.128464     0.675344  
75%                  0.645234          0.163959     0.744324  
max                  0.965300          0.297331     0.969350

## Get accounting returns data.

In [25]:
# returns_url = "http://www.stern.nyu.edu/~adamodar/pc/datasets/pbvdata.xls"
returns_url = "https://github.com/FinancialMarkets/industry_instown_with_data_error_to_clean/blob/master/pbvdata.xls?raw=true"
returns = pd.read_excel(returns_url, skiprows = 7)
returns

Industry  Name  Number of firms       PBV       ROE  \
0                         Advertising               61  5.729582  0.029333   
1                   Aerospace/Defense               72  4.436925  0.085432   
2                       Air Transport               17  3.224150 -0.470269   
3                             Apparel               51  4.111694 -0.081881   
4                        Auto & Truck               19  7.578762  0.044885   
..                                ...              ...       ...       ...   
91                           Trucking               35  4.811726 -0.176956   
92                  Utility (General)               16  1.840435  0.074857   
93                    Utility (Water)               17  3.507385  0.082466   
94                       Total Market             7582  3.813854  0.082466   
95  Total Market (without financials)             6253  4.750108  0.077751   

    EV/ Invested Capital      ROIC  
0               7.009509  0.515119  
1               4.236895  0.191149  
2               1.762879 -0.160654  
3               3.071836  0.075423  
4               2.583233  0.011709  
..                   ...       ...  
91              2.568211 -0.040357  
92              1.476021  0.067851  
93              2.376167  0.060506  
94              2.401274  0.060506  
95              3.423948  0.105844  

[96 rows x 6 columns]

### Merge the two data sets on the Industry Name

In [26]:
all_data = pd.merge(inst_own, returns, on = 'Industry Name')

KeyError: 'Industry Name'

In [27]:
returns.columns

Index(['Industry  Name', 'Number of firms', 'PBV', 'ROE',
       'EV/ Invested Capital', 'ROIC'],
      dtype='object')

In [28]:
inst_own.columns

Index(['Industry Name', 'Number of Firms', 'CEO Holding', 'Corporate Holdings',
       'Institutional Holdings', 'Insider Holdings', 'II_Holdings'],
      dtype='object')

However this throws an error, specifically `KeyError: 'Industry Name'`.  So it is not finding the same column in each `DataFrame`.

With a little inspection, we find that in the returns file, there are two spaces in `Industry  Name` and one space in the institutional ownership file.  

Having spaces in column names is not a good practice, but let's just change the column name in the returns file to have one space between Industry and Name.

In [29]:
returns.rename(columns={'Industry  Name': 'Industry Name'}, inplace=True)

In [30]:
returns.columns

Index(['Industry Name', 'Number of firms', 'PBV', 'ROE',
       'EV/ Invested Capital', 'ROIC'],
      dtype='object')

In [31]:
all_data = pd.merge(inst_own, returns, on = 'Industry Name')

In [32]:
all_data

Industry Name  Number of Firms  CEO Holding  \
0                         Advertising               54     0.099896   
1                   Aerospace/Defense               67     0.034654   
2                       Air Transport               24     0.030976   
3                             Apparel               37     0.053675   
4                        Auto & Truck               34     0.096935   
..                                ...              ...          ...   
90                           Trucking               24     0.054088   
91                  Utility (General)               14     0.001355   
92                    Utility (Water)               15     0.005923   
93                       Total Market             6062     0.048202   
94  Total Market (without financials)             4935     0.048431   

    Corporate Holdings  Institutional Holdings  Insider Holdings  II_Holdings  \
0             0.152666                0.352435          0.233282     0.585717   
1             0.180971                0.528903          0.088726     0.617629   
2             0.261742                0.418113          0.137485     0.555598   
3             0.108202                0.514791          0.170123     0.684914   
4             0.195019                0.230086          0.174605     0.404692   
..                 ...                     ...               ...          ...   
90            0.340488                0.583142          0.168315     0.751457   
91            0.000000                0.857136          0.004635     0.861771   
92            0.276800                0.556169          0.059443     0.615613   
93            0.141806                0.484652          0.127259     0.611911   
94            0.141046                0.496171          0.126508     0.622678   

    Number of firms       PBV       ROE  EV/ Invested Capital      ROIC  
0                61  5.729582  0.029333              7.009509  0.515119  
1                72  4.436925  0.085432              4.236895  0.191149  
2                17  3.224150 -0.470269              1.762879 -0.160654  
3                51  4.111694 -0.081881              3.071836  0.075423  
4                19  7.578762  0.044885              2.583233  0.011709  
..              ...       ...       ...                   ...       ...  
90               35  4.811726 -0.176956              2.568211 -0.040357  
91               16  1.840435  0.074857              1.476021  0.067851  
92               17  3.507385  0.082466              2.376167  0.060506  
93             7582  3.813854  0.082466              2.401274  0.060506  
94             6253  4.750108  0.077751              3.423948  0.105844  

[95 rows x 12 columns]

# Exercise:  

Select only the columns you want and calculate a correlation matrix.  Calculate the matrix.  What is the correlation coefficient between Institutional Ownership and ROE?

In [37]:
all_data

Industry Name  Number of Firms  CEO Holding  \
0                         Advertising               54     0.099896   
1                   Aerospace/Defense               67     0.034654   
2                       Air Transport               24     0.030976   
3                             Apparel               37     0.053675   
4                        Auto & Truck               34     0.096935   
..                                ...              ...          ...   
90                           Trucking               24     0.054088   
91                  Utility (General)               14     0.001355   
92                    Utility (Water)               15     0.005923   
93                       Total Market             6062     0.048202   
94  Total Market (without financials)             4935     0.048431   

    Corporate Holdings  Institutional Holdings  Insider Holdings  II_Holdings  \
0             0.152666                0.352435          0.233282     0.585717   
1             0.180971                0.528903          0.088726     0.617629   
2             0.261742                0.418113          0.137485     0.555598   
3             0.108202                0.514791          0.170123     0.684914   
4             0.195019                0.230086          0.174605     0.404692   
..                 ...                     ...               ...          ...   
90            0.340488                0.583142          0.168315     0.751457   
91            0.000000                0.857136          0.004635     0.861771   
92            0.276800                0.556169          0.059443     0.615613   
93            0.141806                0.484652          0.127259     0.611911   
94            0.141046                0.496171          0.126508     0.622678   

    Number of firms       PBV       ROE  EV/ Invested Capital      ROIC  
0                61  5.729582  0.029333              7.009509  0.515119  
1                72  4.436925  0.085432              4.236895  0.191149  
2                17  3.224150 -0.470269              1.762879 -0.160654  
3                51  4.111694 -0.081881              3.071836  0.075423  
4                19  7.578762  0.044885              2.583233  0.011709  
..              ...       ...       ...                   ...       ...  
90               35  4.811726 -0.176956              2.568211 -0.040357  
91               16  1.840435  0.074857              1.476021  0.067851  
92               17  3.507385  0.082466              2.376167  0.060506  
93             7582  3.813854  0.082466              2.401274  0.060506  
94             6253  4.750108  0.077751              3.423948  0.105844  

[95 rows x 12 columns]

In [40]:
all_data = pd.merge(inst_own, returns, on='Industry Name')
selected_columns = ['Institutional Holdings', 'ROE']
selected_data = all_data[selected_columns]
selected_data


Institutional Holdings       ROE
0                 0.352435  0.029333
1                 0.528903  0.085432
2                 0.418113 -0.470269
3                 0.514791 -0.081881
4                 0.230086  0.044885
..                     ...       ...
90                0.583142 -0.176956
91                0.857136  0.074857
92                0.556169  0.082466
93                0.484652  0.082466
94                0.496171  0.077751

[95 rows x 2 columns]

In [35]:
correlation_matrix = selected_data.corr()
correlation_matrix
print(correlation_matrix)

                        Institutional Holdings       ROE
Institutional Holdings                1.000000  0.077592
ROE                                   0.077592  1.000000


In [36]:
correlation_roe_institutional = correlation_matrix.loc['Institutional Holdings', 'ROE']
print(f"Correlation coefficient between Institutional Ownership and ROE: {correlation_roe_institutional}")

Correlation coefficient between Institutional Ownership and ROE: 0.07759219693460978


#Interpretation:

The correlation coefficient between Institutional Ownership and Return on Equity (ROE) is 0.078. This reflects a weak positive correlation. However, since the result is a positive correlation this suggests that as Institutional Ownership increases, ROE may also increase. Nevertheless, since the value is close to zero, this relationship is not strong or significant.

[Hint](https://pandas.pydata.org/pandas-docs/stable/user_guide/computation.html?highlight=correlation%20matrix)

# Addendum

At this point you may have noticed we call certain 'functions' with `name(a)` and others with `a.name()` where `a` is some object.  As examples, we have used `dir(module)` and `len(object)` but also use `inst_own.sort_values()` and `inst_own.drop()` above.

The difference has to do with the distinction between functions and methods, and this is a topic within Python's class system which we don't want or need to worry about at this point. That said, we use `name(a)` when `name` is a globally defined function.  We use `a.name()` when `name` is a method defined for an object of class (a is an instance of an object of the class).  

You can read more in Section 9.3.4 here: https://docs.python.org/3/tutorial/classes.html

You can read more here and at the referenced links: https://stackoverflow.com/questions/28703834/why-do-some-methods-use-dot-notation-and-others-dont